pkg/log/log.go

```diff
@@ -46,23 +48,50 @@ func WithContext(ctx context.Context) {
 	Info().WithContext(ctx)
 }
 
+func WithError(err error) {
+	Error().WithError(err)
+}
+
 func (l *logger) Print(msg string) {
+	logFields := make(logrus.Fields)
+	if ctxFields := field.FromContext(l.ctx); ctxFields != nil {
+		for _, cf := range ctxFields.Fields() {
+			logFields[cf.Key()] = cf.Value()
+		}
+	}
+
+	if l.entry != nil {
+		l.entry = l.entry.WithFields(logFields)
+	} else {
+		l.entry = logrus.WithFields(logFields)
+	}
+
 	switch l.level {
 	case InfoLevel:
-		logrus.Info(msg)
+		l.entry.Info(msg)
 	case ErrorLevel:
-		logrus.Error(msg)
+		l.entry.Error(msg)
 	case DebugLevel:
-		logrus.Debug(msg)
+		l.entry.Debug(msg)
 	}
 }
 
 func (l *logger) WithContext(ctx context.Context) Logger {
 	l.ctx = ctx
+	if l.entry != nil {
```

Comment:
Later we can figure out an idiomatic way of avoiding having to always check whether entry is `nil` or not.
And in particular, it is desirable to make "logger" immutable. That's mostly implementation and would be compatible with the API currently exposed.